Feature: K8s Deployment Enhancements for Watchdog

sidharth-chauhan opened the pull request:
### Pull Request Description

This pull request introduces Kubernetes deployment configurations for the Watchdog application and updates the project documentation. The following changes have been made:

#### Key Contributions

1. **Kubernetes Manifests Added**:
   - **Deployment Configuration**: 
     - A `deployment.yaml` file has been created to define the deployment of the Watchdog application. This configuration specifies the number of replicas, container image, and ports, ensuring that the application can scale effectively in a Kubernetes environment. The deployment also includes resource limits to optimize performance and manage resource allocation efficiently.
   - **Service Configuration**: 
     - A `service.yaml` file has been added to expose the Watchdog application as a `ClusterIP` service, facilitating internal communication within the Kubernetes cluster.
   - **ConfigMap**: 
     - A `configmap.yaml` file has been created to manage configuration settings for the Watchdog application, including OBA server configurations and other necessary parameters.

2. **Updated README**: 
   - The `README.md` file has been enhanced to provide comprehensive instructions for deploying the Watchdog application using Kubernetes. This includes:
     - **Prerequisites**: Clear requirements for setting up the environment, including necessary tools like Docker, Kind, and kubectl.
     - **Step-by-Step Instructions**: Detailed guidance on creating a Kubernetes cluster using Kind, applying the Kubernetes manifests, and verifying the deployment status. Here are the key steps included in the README:
       - **Create a Kind Cluster**:
         ```bash
         kind create cluster --name watchdog-kind
         ```
       - **Create a Namespace**:
         ```bash
         kubectl create namespace watchdog
         ```
       - **Apply Kubernetes Manifests**:
         ```bash
         kubectl apply -f k8s/ -n watchdog
         ```
       - **Verify Deployment**:
         ```bash
         kubectl get pods -n watchdog
         ```
       - **Access the Service**: The Watchdog service is exposed as a `ClusterIP` service on port `8080`. If the service type is `ClusterIP`, you can use `kubectl port-forward` to access the service locally:
         ```bash
         kubectl port-forward svc/watchdog 8080:8080 -n watchdog
         ```
    

#### Importance of Changes
These enhancements are crucial for improving the deployment capabilities of the Watchdog application. By leveraging Kubernetes, the application can be managed more effectively in a cloud-native environment, allowing for better scalability, reliability, and maintainability. This setup not only simplifies the deployment process but also aligns with modern best practices for application deployment in containerized environments.

<img width="744" alt="Screenshot 2025-04-07 at 2 56 03 AM" src="https://github.com/user-attachments/assets/dbfd9407-7423-4c29-96c1-ec0164bf9aa6" />


### Conclusion
This PR aims to enhance the deployment capabilities of the Watchdog application by leveraging Kubernetes, making it easier to manage and scale the application in a cloud-native environment. Your feedback and suggestions for further improvements are welcome!

sidharth-chauhan commented:
Hi Team,

I have submitted a pull request [Feature: K8s Deployment Enhancements for Watchdog #54](https://github.com/OneBusAway/watchdog/pull/54) that introduces Kubernetes deployment configurations for the Watchdog application. The key contributions include:

1. **Kubernetes Manifests Added:**
   - **Deployment Configuration:** Specifies replicas, container image, and ports.
   - **Service Configuration:** Exposes the application as a ClusterIP service.
   - **ConfigMap:** Manages configuration settings.

2. **Documentation Updates:** Updated the README.md with deployment instructions.

I would appreciate it if you could review the changes and let me know if there are any additional requirements or improvements needed. Please also approve the pending workflow if everything looks good.

aaronbrethorst commented:
thanks but I don't have any way to validate or test this, so I think I need to pass for now

sidharth-chauhan commented:
Thank you for reviewing my previous pull request. I am working on a new pull request addressing the issues identified in PR #54 . I will make sure that the necessary fixes are implemented and will submit the new pull request soon. Your feedback is greatly appreciated!
